pkg/protocol/engine/ledger/ledger/vm.go

```diff
@@ -24,53 +24,82 @@ func (l *Ledger) executeStardustVM(_ context.Context, stateTransition mempool.Tr
-			AccountID:    inp.AccountID,
-			CommitmentID: inp.CommitmentID,
-			Value:        b.Credits.Value,
-		}
+
+	rewardInputs, err := tx.RewardInputs()
+	if err != nil {
+		return nil, xerrors.Errorf("could not get Reward inputs: %w", err)
 	}
 
 	// resolve the commitment inputs from storage
-	commitmentInputSet := iotagovm.CommitmentInputSet{}
-	commitmentInputs, err := tx.CommitmentInputs()
-	if err != nil {
-		return nil, xerrors.Errorf("could not get Commitment inputs: %w", err)
+	commitment := tx.CommitmentInput()
+
+	if (len(rewardInputs) > 0 || len(bicInputs) > 0) && commitment == nil {
+		return nil, xerrors.Errorf("commitment input required with Reward or BIC input")
 	}
-	for _, inp := range commitmentInputs {
-		c, err := l.loadCommitment(inp.CommitmentID)
+
+	var loadedCommitment *iotago.Commitment
+	if commitment != nil {
+		loadedCommitment, err = l.loadCommitment(commitment.CommitmentID)
 		if err != nil {
 			return nil, xerrors.Errorf("could not load commitment: %w", err)
 		}
-		commitmentInputSet[inp.CommitmentID] = c
+
+		resolvedInputs.CommitmentInput = loadedCommitment
 	}
 
-	resolvedInputs := iotagovm.ResolvedInputs{
-		InputSet:           inputSet,
-		BICInputSet:        bicInputSet,
-		CommitmentInputSet: commitmentInputSet,
+	bicInputSet := make(iotagovm.BICInputSet)
+	for _, inp := range bicInputs {
+		accountData, exists, accountErr := l.accountsLedger.Account(inp.AccountID, loadedCommitment.Index)
+		if accountErr != nil {
+			return nil, xerrors.Errorf("could not get BIC input for account %s in slot %d: %w", inp.AccountID, loadedCommitment.Index, accountErr)
+		}
+		if !exists {
+			return nil, xerrors.Errorf("BIC input does not exist for account %s in slot %d", inp.AccountID, loadedCommitment.Index)
+		}
+
+		bicInputSet[inp.AccountID] = accountData.Credits.Value
+	}
+	resolvedInputs.BICInputSet = bicInputSet
+
+	rewardInputSet := make(iotagovm.RewardsInputSet)
+	// when refactoring this to be resolved by the mempool, the function that resolves ContextInputs should receive a slice with promises of UTXO inputs and wait until the necessary
+	for _, inp := range rewardInputs {
+		switch castOutput := inputStates[inp.Index].Output().(type) {
+		case *iotago.AccountOutput:
+			stakingFeature := castOutput.FeatureSet().Staking()
+			if stakingFeature == nil {
+				return nil, xerrors.Errorf("cannot claim rewards from an AccountOutput at index %d without staking feature", inp.Index)
+			}
+			reward, rewardErr := l.epochGadget.ValidatorReward(castOutput.AccountID, stakingFeature.StakedAmount, stakingFeature.StartEpoch, stakingFeature.EndEpoch)
+			if rewardErr != nil {
+				return nil, xerrors.Errorf("failed to get Validator reward for AccountOutput at index %d (StakedAmount: %d, StartEpoch: %d, EndEpoch: %d", castOutput.AccountID, stakingFeature.StakedAmount, stakingFeature.StartEpoch, stakingFeature.EndEpoch)
+			}
+
+			rewardInputSet[castOutput.AccountID] += reward
```

Comment:
```suggestion
			accountID := castOutput.AccountID
			if accountID.Empty() {
				accountID = iotago.AccountIDFromOutputID(inputStates[inp.Index].OutputID())
			}

			reward, rewardErr := l.epochGadget.ValidatorReward(accountID, stakingFeature.StakedAmount, stakingFeature.StartEpoch, stakingFeature.EndEpoch)
			if rewardErr != nil {
				return nil, xerrors.Errorf("failed to get Validator reward for AccountOutput at index %d (StakedAmount: %d, StartEpoch: %d, EndEpoch: %d", accountID, stakingFeature.StakedAmount, stakingFeature.StartEpoch, stakingFeature.EndEpoch)
			}

			rewardInputSet[accountID] = reward
```

- Account ID may be zero, if it is, compute from output ID.
- It should never be legal to include more than one reward input for a given account or delegation ID, but we might not even check for that currently. We should add that check, but it is in any case safer to use `=` here rather than `+=` so we never accidentally pass a multiple of the reward that can actually be claimed.